source/opt/scalar_replacement_pass.cpp

```diff
@@ -0,0 +1,598 @@
+  assert(constant->opcode() == SpvOpConstant ||
+         constant->opcode() == SpvOpConstantNull);
+  if (constant->opcode() == SpvOpConstantNull) {
+    return 0;
+  }
+
+  const ir::Operand& op = constant->GetInOperand(0u);
+  return GetIntegerLiteral(op);
+}
+
+size_t ScalarReplacementPass::GetArrayLength(
+    const ir::Instruction* arrayType) const {
+  assert(arrayType->opcode() == SpvOpTypeArray);
+  const ir::Instruction* length =
+      get_def_use_mgr()->GetDef(arrayType->GetSingleWordInOperand(1u));
+  return GetConstantInteger(length);
+}
+
+size_t ScalarReplacementPass::GetNumElements(
+    const ir::Instruction* type) const {
+  assert(type->opcode() == SpvOpTypeVector ||
+         type->opcode() == SpvOpTypeMatrix);
+  const ir::Operand& op = type->GetInOperand(1u);
+  assert(op.words.size() <= 2);
+  size_t len = 0;
+  for (uint32_t i = 0; i != op.words.size(); ++i) {
+    len |= (op.words[i] << (32 * i));
+  }
+  return len;
+}
+
+ir::Instruction* ScalarReplacementPass::GetStorageType(
+    const ir::Instruction* inst) const {
+  assert(inst->opcode() == SpvOpVariable);
+
+  uint32_t ptrTypeId = inst->type_id();
+  uint32_t typeId =
+      get_def_use_mgr()->GetDef(ptrTypeId)->GetSingleWordInOperand(1u);
+  return get_def_use_mgr()->GetDef(typeId);
+}
+
+bool ScalarReplacementPass::CanReplaceVariable(
+    const ir::Instruction* varInst) const {
+  assert(varInst->opcode() == SpvOpVariable);
+
+  // Can only replace function scope variables.
+  if (varInst->GetSingleWordInOperand(0u) != SpvStorageClassFunction)
+    return false;
+
+  const ir::Instruction* typeInst = GetStorageType(varInst);
+  if (!CheckType(typeInst)) return false;
+
+  if (!CheckUses(varInst)) return false;
+
+  return true;
+}
+
+bool ScalarReplacementPass::CheckType(const ir::Instruction* typeInst) const {
+  switch (typeInst->opcode()) {
+    case SpvOpTypeStruct:
+      // Don't bother with empty structs or very large structs.
+      if (typeInst->NumInOperands() == 0 ||
+          typeInst->NumInOperands() > MAX_NUM_ELEMENTS)
+        return false;
+      return true;
+    case SpvOpTypeArray:
+      if (GetArrayLength(typeInst) > MAX_NUM_ELEMENTS) return false;
+      return true;
+    // Specifically including matrix and vector in an attempt to reduce the
+    // number of vector registers required.
+    case SpvOpTypeMatrix:
+    case SpvOpTypeVector:
+      if (GetNumElements(typeInst) > MAX_NUM_ELEMENTS) return false;
+      return true;
+
+    case SpvOpTypeRuntimeArray:
+    default:
+      return false;
+  }
+}
+
+bool ScalarReplacementPass::CheckUses(const ir::Instruction* inst) const {
+  bool ok = true;
+  VariableStats stats = {0, 0};
+  CheckUses(inst, &stats, &ok);
+
+  // TODO(alanbaker): Extend this to some meaningful heuristics about when SRoA
+  // is valuable.
+  if (stats.num_partial_accesses == 0) ok = false;
+
+  return ok;
+}
+
+void ScalarReplacementPass::CheckUses(const ir::Instruction* inst,
```

Comment:
Why not have it return a bool instead of having |ok| as a parameter?

Reply:
Convenience due to the recursive nature and interaction with ForEachUser lambda.

Reply:
But you can capture &ok in the lambda and return it.